### DISCUSSION TITLE
Lifespan context works only in FastAPI, not in APIRouter

### DISCUSSION THREAD

**Discussion by stur86** [fastapi/fastapi#10464] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from contextlib import asynccontextmanager
from fastapi import APIRouter, FastAPI
from fastapi.testclient import TestClient

values = []

@asynccontextmanager
async def router_lifespan(app: APIRouter) -> None:
    values.append("router_start")
    yield
    values.append("router_end")
    
@asynccontextmanager
async def app_lifespan(app: FastAPI) -> None:
    values.append("app_start")
    yield
    values.append("app_end")

router = APIRouter(lifespan=router_lifespan)
app = FastAPI(lifespan=app_lifespan)
app.include_router(router)

with TestClient(app) as client:
    client.get("/")

print(values)
```

### Description

* I am creating an app with a router, and the router has a `lifespan` method
* I expect the `lifespan` method to handle the lifetime events of the router, same as the old `on_event` method did
* instead, the `lifespan` handler seems to only work for the main app.

### Operating System

Windows

### Operating System Details

Windows 11

### FastAPI Version

0.104.0

### Pydantic Version

2.4.2

### Python Version

3.10.4

**Comment by Kludex** [comment, marked as answer]
- It's not supported yet: https://github.com/encode/starlette/issues/649

**Reply by dzmitry-kankalovich** [reply]
It seems like this is never going to be implemented in Startlette... At this point maybe it's a good idea to deprecate or raise a warning on the `lifespan` property usage, because it's very misleading ATM

**Reply by Kludex** [reply]
We are going to support it, eventually.
